src/text-fragment-utils.js

```diff
@@ -732,49 +732,84 @@ const getBoundaryPointAtIndex = (index, textNodes, isEnd) => {
 
 /**
  * Checks if a substring is word-bounded in the context of a longer string.
- * It's not feasible to match the spec exactly as Intl.Segmenter is not yet
- * widely supported. Instead, returns true iff:
+ *
+ * If an Intl.Segmenter is provided for locale-specific segmenting, it will be
+ * used for this check. This is the most desirable option, but not supported in
+ * all browsers.
+ * 
+ * If one is not provided, a heuristic will be applied,
+ * returning true iff:
  *  - startPos == 0 OR char before start is a boundary char, AND
  *  - length indicates end of string OR char after end is a boundary char
  * Where boundary chars are whitespace/punctuation defined in the const above.
- *
  * This causes the known issue that some languages, notably Japanese, only match
  * at the level of roughly a full clause or sentence, rather than a word.
+ * 
  * @param {String} text - the text to search
  * @param {Number} startPos - the index of the start of the substring
  * @param {Number} length - the length of the substring
+ * @param {Intl.Segmenter} [segmenter] - a segmenter to be used for finding word
+ *     boundaries, if supported
  * @return {bool} - true iff startPos and length point to a word-bounded
  *     substring of |text|.
  */
-const isWordBounded = (text, startPos, length) => {
+const isWordBounded = (text, startPos, length, segmenter) => {
   if (startPos < 0 || startPos >= text.length || length <= 0 ||
       startPos + length > text.length) {
     return false;
   }
 
-  // If the first character is already a boundary, move it once.
-  if (text[startPos].match(BOUNDARY_CHARS)) {
-    ++startPos;
-    --length;
-    if (!length) {
+  if (segmenter) {
+    // If the Intl.Segmenter API is available on this client, use it for more
+    // reliable word boundary checking.
+
+    const segments = segmenter.segment(text);
+    const startSegment = segments.containing(startPos);
+    if (!startSegment) return false;
+    // If the start index is inside a word segment but not the first character
+    // in that segment, it's not word-bounded. If it's not a word segment, then
+    // it's punctuation, etc., so that counts for word bounding.
+    if (startSegment.isWordLike && startSegment.index != startPos) return false;
+
+    const endPos = startPos + length;
+    const endSegment = segments.containing(endPos);
+
+    // If there's no end segment found, it's because we're at the end of the
+    // text, which is a valid boundary. (Because of the preconditions we
+    // checked above, we know we aren't out of range.)
+    // If there is an end segment found, it must either be non-word-like (i.e.,
+    // punctuation/whitespace) or the endPos must indicate the first character
+    // of a new word.
```

Comment:
"or the endPos must indicate the first character o f a new word." Maybe I'm misunderstanding things but to me this doesn't seem like it describes what is actually happening in the check below? The check below seems to say that if `endPos` and `endSegment.index` are the same, then we might be at a word boundary, so it seems like endPos isn't the first character of a new word? Again maybe I'm completely misunderstanding the segmenter API and what `segment.index` means

Reply:
I updated the comment to explain this in more detail. I also added a few test cases--specifically, one that fails if I remove this check, to double-check my work and make sure this really is necessary.

Reply:
Thanks, that helps clarify a bit... But I'm still confused by the fact that we want `endPos` to be at the _start_ of a word... Wouldn't we want `endPos` to be the _end_ of a word to consider the range word-bounded?

Like, if my range is from "H" to "a" in "How are you?", then it shouldn't be considered word bounded, because it ends in the middle of "are". It should include the full "are" to be considered bounded, so we should want `endPos` to be the "e", no?

Reply:
It might be easier to use a different mental model, where you think of the indices as pointing to fenceposts between the characters, rather than the characters themselves:
```
0 1 2 3 4 5 6 7 8 9 A B  <-- char count
|H|o|w| |a|r|e| |y|o|u|  <-- characters
 1 1 1 2 3 3 3 4 5 5 5   <-- segment number
```

Imagine your range is `How ` (note the trailing space). That substring has `startPos` 0, `length` 4, and computed `endPos` 4. Position 4 here is clearly a boundary between segments, so we want to accept it. How do we know it's a boundary between segments? Because segment number 3 starts at position 4; i.e., `segment.index == endPos`. (It would be equally valid to go through and check *every* segment to see if any of them has `index == endPos`, but `segments.containing(endPos)` is a helper that returns the only one that's actually relevant.)

Now a counterexample: imagine the range `How a`. That substring has `startPos` 0, `length` 5, and computed `endPos` 5. In this case, 5 is inside of segment 3, not on a boundary. We know that's not a boundary, because the index of this segment is 4, not 5.